### PR TITLE
Refactors Throw_Impact()

### LINKED ISSUE #5978: You're hit by thrown objects even if you catch them, and here's why.
**Problem Description**:
What is the problem?
You get hit by thrown objects even if you catch them.

**What did you expect to happen**: 
Why do you think this is a bug?
I expected the actions enacted upon me by the thrown object to end at my catching it.

**What happened instead**:
How is what happened different from what you expected?
Once I caught the object, it inflicted damage upon me anyway.

**Why is this bad/What are the consequences:**
Why do you think this is an important issue?
It means that, despite catching thrown objects, they will still damage, embed into and kill players.

**Steps to reproduce the problem**:
The most important section. Review everything you did leading up to causing the issue.
Ensure you've got throw toggled on and get something chucked at you.

**Possibly related stuff (which gamemode was it? What were you doing at the time? Was
anything else out of the ordinary happening?)**: Anything else you can tell us.
Gamemode is irrelevant.

**Why does this happen?**
This bug occurs as a result of an object's _Bump()_ proc conflicting with a mob's _hitby()_ proc because of the _throw_at()_ proc. Or maybe it's because of the _Bump()_ proc...
When hit by an object, that object's _Bump()_ proc calls *throw_impact* which then calls _hitby()_ on the mob which causes the mob to catch the thrown object,
Immediately afterward, the _throw_at()_ proc kicks in and calls *throw_impact* again which then calls _hitby()_ on the mob again as well, causing the mob to sustain damage from the item.

**tl;dr:** _The instant a thrown object comes into contact with a mob, the decision making process in the background that determines what happens makes it seem like the mob was hit by two objects, one they catch and one they don't._

**This is shown by the """""debug messages(tm)"""" in the following screenshot.**
_"What the hell, man"_ is displayed when when the _Bump()_ proc is called.
_"Wow, you caught it. What a beast."_ is displayed when the _hitby()_ proc in /code/modules/mob/living/carbon/human/human_defense.dm registers that you caught the object.
_"Why the hell did it continue running..."_ displays when the _hitby()_ proc in /code/modules/mob/living/carbon/human/human_defense.dm registers that you didn't catch the thrown object, and that it'll do damage to the mob that it hit.

throwhitlogs

### LINKED ISSUE #6201: Chat spam on hurlin' Water Balloons
**Problem Description**:
What is the problem?
On throwing a water balloon it hit a dozen fuckin' times with just one thrown balloon

**What did you expect to happen**: 
Why do you think this is a bug?
It probably should only splat one dude one time, not one dude twenty times.

**What happened instead**:
How is what happened different from what you expected?
It splatted that dude like fifty times.

**Why is this bad/What are the consequences:**
Why do you think this is an important issue?
Low priority but obnoxious chat spam.

**Steps to reproduce the problem**:
The most important section. Review everything you did leading up to causing the issue.
Acquire a water balloon and chuck it at someone. Behold, the sea of "The The Water Balloon bursts!"

**Possibly related stuff (which gamemode was it? What were you doing at the time? Was
anything else out of the ordinary happening?)**: Anything else you can tell us.
Nothing else was particularly special really. I got a water balloon and chucked it at the chaplain, and a thousand messages appears.

Screenshot for proofs:  [http://i.imgur.com/oiwrRz1.png](http://i.imgur.com/oiwrRz1.png)

### PR DESCRIPTION
**Pull request by VampyrBytes**
Splits `throw_impact()` into 2 procs - 

`throw_impact()` deals with the object hitting a mob or object
`ground_impact()` deals with the object hitting the ground - doesn't run if the item doesn't end up on a turf (eg caught)

Also adds an `impact()` proc which is called by both for behavior that happens whether it hits a mob/object or the ground - this has protection so as long as an override runs an `if(!..()) return` it will only run the first time it impacts something.

`throwing` is now set to 0 when an object hits a mob or object - the lack of this is what was causing caught objects to both be caught and hit a mob

fixes #6201 
fixes #5978 
stops throw_impact() behavior being duplicated

### CODE REVIEW

**Comment by KasparoVy** [comment]
Bless you! This is fantastic.

**Comment by tigercat2000** [comment]
We'd probably be better off porting /tg/'s throwing stuff...

**Comment by KasparoVy** [comment]
Why not have an intermediary that works in the mean time?

**Comment by VampyrBytes** [comment]
@tigercat2000 

I did look at /tg/'s throwing code, and yes, it's much nicer, but as @Fox-McCloud points out in #5978 it's not just the throwing code that's involved, it would also be movement and bump code, making it a huge job and one I doubt you'll find anyone willing to undertake - I'm certainly not about to

@KasparoVy 

It was late when I commented on #6295 last night, I should have known exactly what was happening as I'd run into the same issue yesterday sorting water balloons - When I read the comment by @Krausus and realised it was the same issue I had yesterday I figured it needed sorting out - having the same proc deal with hitting the mob/object and the floor was likely to cause issues any time `throw_impact()` is overridden

**Comment by KasparoVy** [comment]
Yeah, I'll rewrite it to conform to the changes this PR makes. I think it's much better waiting till this goes through. It works as it stands but I'm sure it'll be easier now that the procs are split

**Comment by tigercat2000** [comment]
@KasparoVy Intermediaries can make it more difficult to do it properly later; That said, I am not completely sure how much that applies to this specific case, as I haven't looked too far into it.

**Comment by KasparoVy** [comment]
I feel this would be good to have in the meantime. Certainly better having working throwing/catching instead of leaving it busted for an undeterminable length of time holding out hope for a big port

**Comment by VampyrBytes** [comment]
@tigercat2000 A port would involve tearing out all our throwing code anyway, as /tg/ passes the object to a thrownthing datum in a throwing subsystem datum whereas we deal with the item directly, so this would make no difference to a port

**Comment by Fox-McCloud** [comment]
@VampyrBytes 

The previous iteration of that is most likely what we'd go with, as opposed to the subsystem version.

**Comment by VampyrBytes** [comment]
Is it worth me fixing the merge conflicts on this? I'll happily sort them if it's likely to go through, but if it's not going to be accepted I'd rather not bother.

BTW, I've looked again at the core of \TG\'s throw code, and it already has 2 things that I've put in this, namely that throwing is set to 0 when something is hit, and the var to make sure something thrown either hits a mob/object or the ground rather than both. The only part that isn't in \TG\ is the split proc, and that is more for cleanliness than anything as it saves every child proc that needs to know having to have a check to see if it's hit something or just the ground (I've actually realised now I can make the children cleaner - as throw_impact() is now only dealing with 2 situations rather than 3 I can return 1 for a mob hit and 0 for an object hit which will get rid of more child checks - again, will do if this is likely to be accepted)

**Comment by VampyrBytes** [comment]
@Fox-McCloud @tigercat2000 ?

**Comment by Fox-McCloud** [comment]
@VampyrBytes 

I'll try to get this discussed with @Crazylemon64 and @tigercat2000 as quickly as possible.

**Comment by Lady-Luck** [comment]
Does this mean I'll stop taking 10 brute damage from small insignificant items such as ID cards?

**Comment by Fox-McCloud** [comment]
@VampyrBytes 

I haven't forgotten about this--been busy of late.

**Comment by KasparoVy** [comment]
Small conflicts

**Comment by Fox-McCloud** [comment]
Closing this. SOooo sorry for getting back to you on this so late; all our coders have been super super busy of late--we're going to be overhauling the entirety of our `throw_impact` code (and more) to be in line with TG.
